File: CS/Reporting_how-to-create-a-label-in-the-pageheader-band-when-a-field-is-dragged-from-the-t457241/T457241/CustomFieldListDragDropService.cs
using DevExpress.Data.Browsing;
using DevExpress.XtraPrinting;
using DevExpress.XtraReports.Design;
using DevExpress.XtraReports.Expressions.Native;
using DevExpress.XtraReports.UI;
using DevExpress.XtraReports.UserDesigner;
using System;
using System.ComponentModel.Design;
using System.Drawing;
using System.Linq;
using System.Windows.Forms;

namespace T457241 {
    public class CustomFieldListDragDropService : IFieldListDragDropService {
        private IDesignerHost host;
        private XRDesignPanel panel;

        public CustomFieldListDragDropService(IDesignerHost host, XRDesignPanel panel) {
            this.host = host;
            this.panel = panel;
        }
        public IDragHandler GetDragHandler() {
            return new CustomFieldDragHandler(host, panel);
        }
    }

    public class CustomFieldDragHandler : FieldDragHandler {
        XRDesignPanel panel;
        XRControl droppedControl;

        public CustomFieldDragHandler(IDesignerHost host, XRDesignPanel panel)
            : base(host) {
            this.panel = panel;
        }

        public override void HandleDragDrop(object sender, DragEventArgs e) {
            DataInfo[] droppedData = e.Data.GetData(typeof(DataInfo[])) as DataInfo[];
            XRControl parentControl = BandViewSvc.GetControlByScreenPoint(new Point(e.X, e.Y));

            ISelectionService selectSvc = Host.GetService(typeof(ISelectionService)) as ISelectionService;

            if(((parentControl is XRPanel) || (parentControl is Band)) && ((droppedData.Length == 1))) {
                AddSingleField(e, droppedData, parentControl, selectSvc);
            } else if(((parentControl is XRPanel) || (parentControl is Band)) && ((droppedData.Length > 1)))
                AddMultipleFields(e, droppedData, parentControl, selectSvc);
            else
                base.HandleDragDrop(sender, e);
        }

        private void AddMultipleFields(DragEventArgs e, DataInfo[] droppedData, XRControl parentControl, ISelectionService selectSvc) {
            this.AdornerService.ResetAll();
            this.RulerService.HideShadows();
            XRTableCell headerCell;
            XRControl parent = BandViewSvc.GetControlByScreenPoint(new Point(e.X, e.Y));
            if(parent == null)
                return;
            SizeF size = new SizeF(100F * droppedData.Length, 25F);

            if(parentControl is DetailBand)
                size.Width = CalculateWidth(parentControl);

            XRTable detailTable = new XRTable() { Name = "DetailTable" };
            detailTable.BeginInit();
            XRTableRow detailRow = new XRTableRow();
            detailTable.Rows.Add(detailRow);

            this.droppedControl = detailTable;
            detailTable.SizeF = size;

            Host.Container.Add(detailTable);
            Host.Container.Add(detailRow);

            for(int i = 0; i < droppedData.Length; i++) {
                XRTableCell cell = new XRTableCell();
                string expression = ExpressionBindingHelper.NormalizeDataMember(droppedData[i].Member, parent.Report.DataMember);
                cell.ExpressionBindings.Add(
                    new ExpressionBinding("Text",
                    expression));
                detailRow.Cells.Add(cell);
                Host.Container.Add(cell);
            }
            detailTable.EndInit();           

            PointF dropPoint = GetDragDropLocation(e, detailTable, parentControl);
            this.DropXRControl(parentControl, new PointF(0, dropPoint.Y));
            selectSvc.SetSelectedComponents(new XRControl[] { detailTable });
            if((parentControl is DetailBand)) {
                PageHeaderBand band = null;
                if((parentControl as DetailBand).Report.Bands.OfType<PageHeaderBand>().FirstOrDefault() != null) {
                    band = (parentControl as DetailBand).Report.Bands[BandKind.PageHeader] as PageHeaderBand;
                } else {
                    band = new PageHeaderBand();
                    (parentControl as DetailBand).Report.Bands.Add(band);
                    Host.Container.Add(band);
                }

                XRTable headerTable = new XRTable() { Name = "HeaderTable" };
                headerTable.BeginInit();

                XRTableRow headerRow = new XRTableRow();
                headerTable.Rows.Add(headerRow);

                headerTable.SizeF = size;

                Host.Container.Add(headerTable);
                Host.Container.Add(headerRow);

                for(int i = 0; i < droppedData.Length; i++)
                    headerCell = CreateTableCell(Host, headerRow, droppedData[i].DisplayName);
                headerTable.Borders = BorderSide.All;
                headerTable.EndInit();

                band.Controls.Add(headerTable);
                headerTable.LocationF = droppedControl.LocationF;
            }
        }

        XRTableCell CreateTableCell(IDesignerHost host, XRTableRow row, string cellText) {
            XRTableCell headerCell = new XRTableCell();
            headerCell.Text = cellText;
            headerCell.BackColor = Color.Green;
            headerCell.ForeColor = Color.Yellow;
            headerCell.TextAlignment = DevExpress.XtraPrinting.TextAlignment.MiddleCenter;
            headerCell.Font = new Font("Calibry", 11, FontStyle.Bold);
            row.Cells.Add(headerCell);
            host.Container.Add(headerCell);
            return headerCell;
        }

        private PointF GetDragDropLocation(DragEventArgs e, XRControl control, XRControl parent) {
            PointF bandPoint = EvalBandPoint(e, parent.Band);
            bandPoint = BandViewSvc.SnapBandPoint(bandPoint, parent.Band, control, new XRControl[] { control });
            PointF screenPoint = BandViewSvc.ControlViewToScreen(bandPoint, parent.Band);
            return BandViewSvc.ScreenToControl(new RectangleF(screenPoint, SizeF.Empty), parent).Location;
        }

        private float CalculateWidth(XRControl control) {
            XtraReport report = control.RootReport;
            return GraphicsUnitConverter.Convert(report.PageWidth - report.Margins.Left - report.Margins.Right, report.Dpi, GraphicsDpi.HundredthsOfAnInch);
        }

        private void AddSingleField(DragEventArgs e, DataInfo[] droppedData, XRControl parentControl, ISelectionService selectSvc) {
            this.AdornerService.ResetAll();
            this.RulerService.HideShadows();

            SizeF size = new SizeF(100F, 25F);

            XRLabel detailLabel = new XRLabel();
            this.droppedControl = detailLabel;
            detailLabel.SizeF = size;

            Host.Container.Add(detailLabel);
            PointF dropPoint = GetDragDropLocation(e, detailLabel, parentControl);

            detailLabel.ExpressionBindings.Add(new ExpressionBinding("Text",
                ExpressionBindingHelper.NormalizeDataMember(droppedData[0].Member, parentControl.Report.DataMember)));           

            this.DropXRControl(parentControl, dropPoint);
            selectSvc.SetSelectedComponents(new XRControl[] { detailLabel });
            if((parentControl is DetailBand)) {
                PageHeaderBand band = null;
                if(panel.Report.Bands.OfType<PageHeaderBand>().FirstOrDefault() != null) {
                    band = panel.Report.Bands[BandKind.PageHeader] as PageHeaderBand;
                } else {
                    band = new PageHeaderBand();
                    panel.Report.Bands.Add(band);
                    Host.Container.Add(band);
                }
                XRLabel headerLabel = CreateLabel(droppedControl.LocationF, size, droppedData[0].DisplayName);
                Host.Container.Add(headerLabel);
                band.Controls.Add(headerLabel);
            }
        }

        XRLabel CreateLabel(PointF location, SizeF size, string labelText) {
            XRLabel headerLabel = new XRLabel();
            headerLabel.SizeF = size;
            headerLabel.LocationF = location;
            headerLabel.Text = labelText;
            headerLabel.BackColor = Color.Green;
            headerLabel.ForeColor = Color.Yellow;
            headerLabel.TextAlignment = DevExpress.XtraPrinting.TextAlignment.MiddleCenter;
            headerLabel.Font = new Font("Calibry", 11, FontStyle.Bold);
            return headerLabel;
        }

        void DropXRControl(XRControl parent, PointF dropPoint) {
            PointF screenPoint = dropPoint;
            parent.Controls.Add(droppedControl);
            droppedControl.LocationF = dropPoint;
        }
    }
}

File: VB/Reporting_how-to-create-a-label-in-the-pageheader-band-when-a-field-is-dragged-from-the-t457241/T457241/CustomFieldListDragDropService.vb
Imports DevExpress.Data.Browsing
Imports DevExpress.XtraPrinting
Imports DevExpress.XtraReports.Design
Imports DevExpress.XtraReports.Expressions.Native
Imports DevExpress.XtraReports.UI
Imports DevExpress.XtraReports.UserDesigner
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel.Design
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms

Namespace T457241

    Public Class CustomFieldListDragDropService
        Implements IFieldListDragDropService

        Private host As IDesignerHost
        Private panel As XRDesignPanel

        Public Sub New(ByVal host As IDesignerHost, ByVal panel As XRDesignPanel)
            Me.host = host
            Me.panel = panel
        End Sub

        Public Function GetDragHandler() As IDragHandler Implements IFieldListDragDropService.GetDragHandler
            Return New CustomFieldDragHandler(host, panel)
        End Function
    End Class

    Public Class CustomFieldDragHandler
        Inherits FieldDragHandler

        Private panel As XRDesignPanel
        Private droppedControl As XRControl

        Public Sub New(ByVal host As IDesignerHost, ByVal panel As XRDesignPanel)
            MyBase.New(host)
            Me.panel = panel
        End Sub

        Public Overrides Sub HandleDragDrop(ByVal sender As Object, ByVal e As DragEventArgs)
            ResetUI(e)

            Dim droppedData() As DataInfo = TryCast(e.Data.GetData(GetType(DataInfo())), DataInfo())
            Dim parentControl As XRControl = BandViewSvc.GetControlByScreenPoint(New System.Drawing.Point(e.X, e.Y))

            Dim selectSvc As ISelectionService = TryCast(Host.GetService(GetType(ISelectionService)), ISelectionService)

            If ((TypeOf parentControl Is XRPanel) OrElse (TypeOf parentControl Is Band)) AndAlso ((droppedData.Length = 1)) Then
                AddSingleField(e, droppedData, parentControl, selectSvc)
            ElseIf ((TypeOf parentControl Is XRPanel) OrElse (TypeOf parentControl Is Band)) AndAlso ((droppedData.Length > 1)) Then
                AddMultipleFields(e, droppedData, parentControl, selectSvc)
            Else
                MyBase.HandleDragDrop(sender, e)
            End If
        End Sub

        Private Sub AddMultipleFields(ByVal e As DragEventArgs, ByVal droppedData() As DataInfo, ByVal parentControl As XRControl, ByVal selectSvc As ISelectionService)
            Dim headerCell As XRTableCell
            Dim parent As XRControl = BandViewSvc.GetControlByScreenPoint(New System.Drawing.Point(e.X, e.Y))
            If parent Is Nothing Then
                Return
            End If
            Dim size As New SizeF(100.0F * droppedData.Length, 25.0F)

            If TypeOf parentControl Is DetailBand Then
                size.Width = CalculateWidth(parentControl)
            End If

            Dim detailTable As New XRTable() With {.Name = "DetailTable"}
            detailTable.BeginInit()
            Dim detailRow As New XRTableRow()
            detailTable.Rows.Add(detailRow)

            Me.droppedControl = detailTable
            detailTable.SizeF = size

            Host.Container.Add(detailTable)
            Host.Container.Add(detailRow)

            For i As Integer = 0 To droppedData.Length - 1
                Dim cell As New XRTableCell()
                cell.ExpressionBindings.Add(New ExpressionBinding("Text", ExpressionBindingHelper.NormalizeDataMember(droppedData(i).Member, parentControl.Report.DataMember)))
                detailRow.Cells.Add(cell)
                Host.Container.Add(cell)
            Next i
            detailTable.EndInit()

            Dim dropPoint As PointF = GetDragDropLocation(e, detailTable, parentControl)
            Me.DropXRControl(parentControl, New PointF(0, dropPoint.Y))
            selectSvc.SetSelectedComponents(New XRControl() {detailTable})
            If (TypeOf parentControl Is DetailBand) Then
                Dim band As PageHeaderBand = Nothing
                If (TryCast(parentControl, DetailBand)).Report.Bands.OfType(Of PageHeaderBand)().FirstOrDefault() IsNot Nothing Then
                    band = TryCast((TryCast(parentControl, DetailBand)).Report.Bands(BandKind.PageHeader), PageHeaderBand)
                Else
                    band = New PageHeaderBand()
                    TryCast(parentControl, DetailBand).Report.Bands.Add(band)
                    Host.Container.Add(band)
                End If

                Dim headerTable As New XRTable() With {.Name = "HeaderTable"}
                headerTable.BeginInit()

                Dim headerRow As New XRTableRow()
                headerTable.Rows.Add(headerRow)

                headerTable.SizeF = size

                Host.Container.Add(headerTable)
                Host.Container.Add(headerRow)

                For i As Integer = 0 To droppedData.Length - 1
                    headerCell = CreateTableCell(headerRow, droppedData(i).DisplayName)
                Next i
                headerTable.Borders = BorderSide.All
                headerTable.EndInit()

                band.Controls.Add(headerTable)
                headerTable.LocationF = droppedControl.LocationF
            End If
        End Sub

        Private Function CreateTableCell(ByVal row As XRTableRow, ByVal cellText As String) As XRTableCell
            Dim headerCell As New XRTableCell()
            headerCell.Text = cellText
            headerCell.BackColor = Color.Green
            headerCell.ForeColor = Color.Yellow
            headerCell.TextAlignment = DevExpress.XtraPrinting.TextAlignment.MiddleCenter
            headerCell.Font = New Font("Calibry", 11, FontStyle.Bold)
            row.Cells.Add(headerCell)
            Host.Container.Add(headerCell)
            Return headerCell
        End Function


        Private Function GetDragDropLocation(ByVal e As DragEventArgs, ByVal control As XRControl, ByVal parent As XRControl) As PointF
            Dim bandPoint As PointF = EvalBandPoint(e, parent.Band)
            bandPoint = BandViewSvc.SnapBandPoint(bandPoint, parent.Band, control, New XRControl() {control})
            Dim screenPoint As PointF = BandViewSvc.ControlViewToScreen(bandPoint, parent.Band)
            Return BandViewSvc.ScreenToControl(New RectangleF(screenPoint, SizeF.Empty), parent).Location
        End Function


        Private Function CalculateWidth(ByVal control As XRControl) As Single
            Dim report As XtraReport = control.RootReport
            Return GraphicsUnitConverter.Convert(report.PageWidth - report.Margins.Left - report.Margins.Right, report.Dpi, GraphicsDpi.HundredthsOfAnInch)
        End Function

        Private Sub AddSingleField(ByVal e As DragEventArgs, ByVal droppedData() As DataInfo, ByVal parentControl As XRControl, ByVal selectSvc As ISelectionService)
            Dim size As New SizeF(100.0F, 25.0F)

            Dim detailLabel As New XRLabel()
            Me.droppedControl = detailLabel
            detailLabel.SizeF = size

            Host.Container.Add(detailLabel)
            Dim dropPoint As PointF = GetDragDropLocation(e, detailLabel, parentControl)
            detailLabel.ExpressionBindings.Add(New ExpressionBinding("Text", ExpressionBindingHelper.NormalizeDataMember(droppedData(0).Member, parentControl.Report.DataMember)))

            Me.DropXRControl(parentControl, dropPoint)
            selectSvc.SetSelectedComponents(New XRControl() {detailLabel})
            If (TypeOf parentControl Is DetailBand) Then
                Dim band As PageHeaderBand = Nothing
                If panel.Report.Bands.OfType(Of PageHeaderBand)().FirstOrDefault() IsNot Nothing Then
                    band = TryCast(panel.Report.Bands(BandKind.PageHeader), PageHeaderBand)
                Else
                    band = New PageHeaderBand()
                    panel.Report.Bands.Add(band)
                    Host.Container.Add(band)
                End If
                Dim headerLabel As XRLabel = CreateLabel(droppedControl.LocationF, size, droppedData(0).DisplayName)
                Host.Container.Add(headerLabel)
                band.Controls.Add(headerLabel)
            End If
        End Sub

        Private Function CreateLabel(ByVal location As PointF, ByVal size As SizeF, ByVal labelText As String) As XRLabel
            Dim headerLabel As New XRLabel()
            headerLabel.SizeF = size
            headerLabel.LocationF = location
            headerLabel.Text = labelText
            headerLabel.BackColor = Color.Green
            headerLabel.ForeColor = Color.Yellow
            headerLabel.TextAlignment = DevExpress.XtraPrinting.TextAlignment.MiddleCenter
            headerLabel.Font = New Font("Calibry", 11, FontStyle.Bold)
            Return headerLabel
        End Function

        Private Sub DropXRControl(ByVal parent As XRControl, ByVal dropPoint As PointF)
            parent.Controls.Add(droppedControl)
            droppedControl.LocationF = dropPoint
        End Sub
    End Class
End Namespace
